wasm/src/decoder/common.rs

```diff
@@ -0,0 +1,94 @@
+#[repr(u8)]
+pub enum ExternalKind {
+    Function = 0x00,
+    Table = 0x01,
+    Memory = 0x02,
+    Global = 0x03,
+}
+
+impl From<ExternalKind> for u8 {
+    fn from(v: ExternalKind) -> Self {
+        v as Self
+    }
+}
+
+impl TryFrom<u8> for ExternalKind {
+    type Error = ();
+
+    fn try_from(v: u8) -> Result<Self, Self::Error> {
+        match v {
+            0x00 => Ok(Self::Function),
+            0x01 => Ok(Self::Table),
+            0x02 => Ok(Self::Memory),
+            0x03 => Ok(Self::Global),
+            _ => Err(()),
+        }
+    }
+}
+
+#[repr(u8)]
+#[derive(Copy, Clone)]
+pub enum ValueKind {
+    I32 = 0x7F,
+    I64 = 0x7E,
+    F32 = 0x7D,
+    F64 = 0x7C,
+    Funcref = 0x70,
+    Func = 0x60,
+    Void = 0x40,
+}
+
+impl From<ValueKind> for u8 {
+    fn from(v: ValueKind) -> Self {
+        v as Self
+    }
+}
+
+impl TryFrom<u8> for ValueKind {
+    type Error = ();
+
+    fn try_from(v: u8) -> Result<Self, Self::Error> {
+        match v {
+            0x7F => Ok(Self::I32),
+            0x7E => Ok(Self::I64),
+            0x7D => Ok(Self::F32),
+            0x7C => Ok(Self::F64),
+            0x70 => Ok(Self::Funcref),
+            0x60 => Ok(Self::Func),
+            0x40 => Ok(Self::Void),
+            _ => Err(()),
+        }
+    }
+}
+
+pub struct FnType {
+    pub params: Vec<ValueKind>,
+    pub result: Vec<ValueKind>,
```

Comment:
This is plural, how about `results`?

Reply:
your call

Reply:
In .wat, it's singular "param" though too `(param i32 i32 i32)`, and arguably we should just spell it out for "parameters." I'd say they should both be plural.